adapters/smaato/smaato.go

```diff
@@ -0,0 +1,289 @@
+		if userExtErr != nil {
+			errs = append(errs, userExtErr)
+			return nil, errs
+		}
+
+		userCopy := *request.User
+		extractUserExtAttributes(userExt, &userCopy)
+		delete(userExtRaw, "data")
+		userCopy.Ext, err = json.Marshal(userExtRaw)
+		if err != nil {
+			errs = append(errs, err)
+			return nil, errs
+		}
+		request.User = &userCopy
+	}
+
+	// Setting ext client info
+	type bidRequestExt struct {
+		Client string `json:"client"`
+	}
+	request.Ext, err = json.Marshal(bidRequestExt{Client: clientVersion})
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+	reqJSON, err := json.Marshal(request)
+	if err != nil {
+		errs = append(errs, err)
+		return nil, errs
+	}
+
+	uri := a.URI
+
+	headers := http.Header{}
+	headers.Add("Content-Type", "application/json;charset=utf-8")
+	headers.Add("Accept", "application/json")
+
+	return []*adapters.RequestData{{
+		Method:  "POST",
+		Uri:     uri,
+		Body:    reqJSON,
+		Headers: headers,
+	}}, errs
+}
+
+// MakeBids unpacks the server's response into Bids.
+func (a *SmaatoAdapter) MakeBids(internalRequest *openrtb.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{fmt.Errorf("unexpected status code: %d. Run with request.debug = 1 for more info", response.StatusCode)}
+	}
+
+	var bidResp openrtb.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(5)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := 0; i < len(sb.Bid); i++ {
+			bid := sb.Bid[i]
+
+			var markupError error
+			bid.AdM, markupError = renderAdMarkup(getAdMarkupType(response, bid.AdM), bid.AdM)
+			if markupError != nil {
+				fmt.Println(markupError)
+				continue // no bid when broken ad markup
+			}
+
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: openrtb_ext.BidTypeBanner,
+			})
+		}
+	}
+	return bidResponse, nil
+}
+
+func renderAdMarkup(adMarkupType adMarkupType, adMarkup string) (string, error) {
+	var markupError error
+	var adm string
+	switch adMarkupType {
+	case smtAdTypeImg:
+		adm, markupError = extractAdmImage(adMarkup)
+	case smtAdTypeRichmedia:
+		adm, markupError = extractAdmRichMedia(adMarkup)
+	default:
+		return "", fmt.Errorf("Unknown markup type %s", adMarkupType)
+	}
+	return adm, markupError
+}
+
+func getAdMarkupType(response *adapters.ResponseData, adMarkup string) adMarkupType {
+	admType := adMarkupType(response.Headers.Get("X-SMT-ADTYPE"))
```

Comment:
What do you think of returning immediately instead of continuing checks for an empty string?

```
	if admType := adMarkupType(response.Headers.Get("X-SMT-ADTYPE")); admType != "" {
		return admType
	}

	if strings.HasPrefix(adMarkup, `{"image":`) {
		return smtAdTypeImg
	}

	if strings.HasPrefix(adMarkup, `{"richmedia":`) {
		return smtAdTypeRichmedia
	}

	return ""
```

Reply:
done